Use the official aws-sdk instead of jet3t

jihoonson opened the pull request:
Fixes #4289.

In this PR, the version of the below libraries are changed.

- Hadoop => ~~2.9.0 to match library versions~~ 2.8.3
- AWS-SDK => 1.11.199 (https://github.com/druid-io/druid/issues/4456)
- Jackson => ~~2.8.10~~ 2.6.7
- HttpClient => 4.5.3 (https://github.com/druid-io/druid/blob/master/common/pom.xml#L160-L169, https://issues.apache.org/jira/browse/HTTPCLIENT-1727)

Also, I removed `AWSCredentialsProvider` which is not used anymore.

I've done the below tests.

- Indexing
	- Hadoop indexing with Hadoop 2.7.5 (works with S3NFileSystem)
	- Hadoop indexing with Hadoop 2.8.3
	- Hadoop indexing with Hadoop 2.9.0
	- Hadoop indexing with Hadoop 3.0 (works when `hadoopDependencyCoordinates` was set to 2.9.0)
	- Local indexing
		- With prefetch
		- Without prefetch
		- Multi-region support: can read data from multiple regions
- `insert-segment-to-db` tool: this tool finds segments under the given directory of deep storage
- Moving segments with moveTask
- Killing segments with killTask

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/5382)
<!-- Reviewable:end -->

b-slim commented:
@jihoonson does this means we are dropping support for Hadoop < 2.9.0 ?

jihoonson commented:
@b-slim yes. I tested with 2.7.5, but it didn't work because of the mismatched version of aws-sdk library.

jihoonson commented:
I added `WIP` label because this PR needs more testing on larger clusters. My tests were done only in my laptop.

jihoonson commented:
Also, there are some `TODO` notes I left which should be addressed.

gianm commented:
Is it possible to keep the same version of aws-java-sdk as we have now, and do this patch without changing the Hadoop version? Or is there a reason they both need to be upgraded?

I am just asking because people struggle with upgrades every time we change the Hadoop version, so I was hoping it wouldn't be necessary (I was hoping the next upgrade after 2.7.3 would be to 3.x, sometime in the future after its deployment becomes more widespread).

jihoonson commented:
@gianm yes, there is a reason. To use aws-java-sdk, we need to match its version to that used by Hadoop as well as other libraries' versions like jackson, httpclient, and so on. (Unfortunately, I lost my note listing the libraries needing version change.) From my testing, 2.9.0 was the oldest working version.

gianm commented:
> @gianm yes, there is a reason. To use aws-java-sdk, we need to match its version to that used by Hadoop as well as other libraries' versions like jackson, httpclient, and so on. (Unfortunately, I lost my note listing the libraries needing version change.) From my testing, 2.9.0 was the oldest working version.

I don't understand why we need to use aws-java-sdk from Hadoop at all. I thought that when we run Hadoop jobs, we use its FileSystem implementations both for reading (possibly from S3) and writing (to a possibly S3 deep storage) and do not call jets3t directly. Could we solve this by excluding the java sdk from the main classloader when we are running Hadoop jobs, meaning we only load the one provided in `hadoop-dependencies`?

gianm commented:
> I don't understand why we need to use aws-java-sdk from Hadoop at all. 

This should say: I don't understand why we need to use _Druid's version of_ aws-java-sdk from Hadoop at all.

b-slim commented:
FYI am all in to move to 3.0 but I really think we should be able to support hadoop 2.7 line (if possible) 2.9 is not really popular and I don't think it will be, the most likely, the users will stick with 2.7 line or do the leap and move to 3.0.  
@jihoonson can you please elaborate around https://github.com/druid-io/druid/pull/5382#issuecomment-365062725 ? what is the test that failed and what kind of dependency is breaking? Thanks

jihoonson commented:
> I don't understand why we need to use aws-java-sdk from Hadoop at all. I thought that when we run Hadoop jobs, we use its FileSystem implementations both for reading (possibly from S3) and writing (to a possibly S3 deep storage) and do not call jets3t directly.

Yes, this is correct. What I meant is the old implementation of S3AFileSystem uses some deprecated APIs of aws-sdk which are not in the recent versions. This is not just an issue of aws-sdk, but also some other libraries like jackson.

jihoonson commented:
@b-slim when I tested with hadoop 2.7.5, I could see the below error.

```
2018-02-13T05:02:23,061 ERROR [task-runner-0-priority-0] io.druid.indexing.overlord.ThreadPoolTaskRunner - Exception while running task[HadoopIndexTask{id=index_hadoop_lineitem_2018-02-13T05:02:00.954Z, type=index_hadoop, dataSource=lineitem}]
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[guava-16.0.1.jar:?]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:222) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.common.task.HadoopIndexTask.run(HadoopIndexTask.java:184) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:450) [druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:422) [druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_161]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_161]
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:219) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
Caused by: java.lang.NoSuchMethodError: com.amazonaws.services.s3.transfer.TransferManager.<init>(Lcom/amazonaws/services/s3/AmazonS3;Ljava/util/concurrent/ThreadPoolExecutor;)V
	at org.apache.hadoop.fs.s3a.S3AFileSystem.initialize(S3AFileSystem.java:287) ~[?:?]
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2667) ~[?:?]
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:93) ~[?:?]
	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2701) ~[?:?]
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2683) ~[?:?]
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:372) ~[?:?]
	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:295) ~[?:?]
	at org.apache.hadoop.mapreduce.lib.input.FileInputFormat.setInputPaths(FileInputFormat.java:500) ~[?:?]
	at org.apache.hadoop.mapreduce.lib.input.DelegatingInputFormat.getSplits(DelegatingInputFormat.java:110) ~[?:?]
	at org.apache.hadoop.mapreduce.JobSubmitter.writeNewSplits(JobSubmitter.java:301) ~[?:?]
	at org.apache.hadoop.mapreduce.JobSubmitter.writeSplits(JobSubmitter.java:318) ~[?:?]
	at org.apache.hadoop.mapreduce.JobSubmitter.submitJobInternal(JobSubmitter.java:196) ~[?:?]
	at org.apache.hadoop.mapreduce.Job$10.run(Job.java:1290) ~[?:?]
	at org.apache.hadoop.mapreduce.Job$10.run(Job.java:1287) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_161]
	at javax.security.auth.Subject.doAs(Subject.java:422) ~[?:1.8.0_161]
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1754) ~[?:?]
	at org.apache.hadoop.mapreduce.Job.submit(Job.java:1287) ~[?:?]
	at io.druid.indexer.DetermineHashedPartitionsJob.run(DetermineHashedPartitionsJob.java:119) ~[druid-indexing-hadoop-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexer.JobHelper.runJobs(JobHelper.java:362) ~[druid-indexing-hadoop-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexer.HadoopDruidDetermineConfigurationJob.run(HadoopDruidDetermineConfigurationJob.java:91) ~[druid-indexing-hadoop-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.common.task.HadoopIndexTask$HadoopDetermineConfigInnerProcessing.runTask(HadoopIndexTask.java:325) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:219) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
```

This error means that S3AFileSystem of 2.7.5 is using a constructor of TransferManager which is not in the recent version of aws-sdk. So, I added an older version (1.7.4) of aws-sdk to the classpath, and changed the dependency order in the classpath for aws-sdk 1.7.4 to be appeared earlier. This caused another error shown below.

```
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Container killed by the ApplicationMaster.
Container killed on request. Exit code is 143
Container exited with a non-zero exit code 143

Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
Error: class com.fasterxml.jackson.datatype.guava.deser.HostAndPortDeserializer overrides final method deserialize.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/databind/DeserializationContext;)Ljava/lang/Object;
2018-02-12T23:37:31,308 ERROR [task-runner-0-priority-0] io.druid.indexing.overlord.ThreadPoolTaskRunner - Exception while running task[HadoopIndexTask{id=index_hadoop_lineitem_2018-02-12T23:36:37.550Z, type=index_hadoop, dataSource=lineitem}]
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[guava-16.0.1.jar:?]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:231) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.common.task.HadoopIndexTask.run(HadoopIndexTask.java:184) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:450) [druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:422) [druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_161]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_161]
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:228) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
Caused by: io.druid.java.util.common.ISE: Job[class io.druid.indexer.DetermineHashedPartitionsJob] failed!
	at io.druid.indexer.JobHelper.runJobs(JobHelper.java:384) ~[druid-indexing-hadoop-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexer.HadoopDruidDetermineConfigurationJob.run(HadoopDruidDetermineConfigurationJob.java:91) ~[druid-indexing-hadoop-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.common.task.HadoopIndexTask$HadoopDetermineConfigInnerProcessing.runTask(HadoopIndexTask.java:325) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:228) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
```

This error came from the mismatched version of Jackson, so I also added an older version of Jackson to the classpath as well and adjusted the order. This causes the below error.

```
2018-02-12T23:56:18,757 ERROR [task-runner-0-priority-0] io.druid.indexing.overlord.ThreadPoolTaskRunner - Exception while running task[HadoopIndexTask{id=index_hadoop_lineitem_2018-02-12T23:56:12.573Z, type=index_hadoop, dataSource=lineitem}]
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[guava-16.0.1.jar:?]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:231) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.common.task.HadoopIndexTask.run(HadoopIndexTask.java:184) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:450) [druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:422) [druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_161]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_161]
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:228) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
Caused by: java.lang.ExceptionInInitializerError
	at io.druid.indexing.common.task.HadoopIndexTask$HadoopDetermineConfigInnerProcessing.runTask(HadoopIndexTask.java:311) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:228) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
Caused by: com.google.inject.CreationException: Unable to create injector, see the following errors:

1) Error in custom provider, java.lang.VerifyError: Cannot inherit from final class
  at io.druid.jackson.JacksonModule.jsonMapper(JacksonModule.java:46)
  at io.druid.jackson.JacksonModule.jsonMapper(JacksonModule.java:46)
  while locating com.fasterxml.jackson.databind.ObjectMapper annotated with interface io.druid.guice.annotations.Json
  while locating com.fasterxml.jackson.databind.ObjectMapper
    for the 1st parameter of io.druid.guice.JsonConfigurator.<init>(JsonConfigurator.java:67)
  at io.druid.guice.ConfigModule.configure(ConfigModule.java:40)
  while locating io.druid.guice.JsonConfigurator
    for the 2nd parameter of io.druid.guice.JsonConfigProvider.inject(JsonConfigProvider.java:188)
  at io.druid.guice.JsonConfigProvider.bind(JsonConfigProvider.java:133)

2) Error in custom provider, java.lang.VerifyError: Cannot inherit from final class
  at io.druid.jackson.JacksonModule.jsonMapper(JacksonModule.java:46)
  at io.druid.jackson.JacksonModule.jsonMapper(JacksonModule.java:46)
  while locating com.fasterxml.jackson.databind.ObjectMapper annotated with interface io.druid.guice.annotations.Json
  while locating com.fasterxml.jackson.databind.ObjectMapper
    for the 1st parameter of io.druid.guice.JsonConfigurator.<init>(JsonConfigurator.java:67)
  at io.druid.guice.ConfigModule.configure(ConfigModule.java:40)
  while locating io.druid.guice.JsonConfigurator
    for the 2nd parameter of io.druid.guice.JsonConfigProvider.inject(JsonConfigProvider.java:188)
  at io.druid.guice.JsonConfigProvider.bind(JsonConfigProvider.java:133)

2 errors
	at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:470) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:176) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:110) ~[guice-4.1.0.jar:?]
	at com.google.inject.Guice.createInjector(Guice.java:99) ~[guice-4.1.0.jar:?]
	at com.google.inject.Guice.createInjector(Guice.java:73) ~[guice-4.1.0.jar:?]
	at io.druid.guice.GuiceInjectors.makeStartupInjector(GuiceInjectors.java:60) ~[druid-processing-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexer.HadoopDruidIndexerConfig.<clinit>(HadoopDruidIndexerConfig.java:106) ~[druid-indexing-hadoop-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.common.task.HadoopIndexTask$HadoopDetermineConfigInnerProcessing.runTask(HadoopIndexTask.java:311) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:228) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
Caused by: java.lang.VerifyError: Cannot inherit from final class
	at java.lang.ClassLoader.defineClass1(Native Method) ~[?:1.8.0_161]
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763) ~[?:1.8.0_161]
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142) ~[?:1.8.0_161]
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467) ~[?:1.8.0_161]
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73) ~[?:1.8.0_161]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368) ~[?:1.8.0_161]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362) ~[?:1.8.0_161]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_161]
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361) ~[?:1.8.0_161]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_161]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_161]
	at com.fasterxml.jackson.datatype.guava.GuavaModule.setupModule(GuavaModule.java:55) ~[jackson-datatype-guava-2.8.10.jar:2.8.10]
	at com.fasterxml.jackson.databind.ObjectMapper.registerModule(ObjectMapper.java:524) ~[jackson-databind-2.8.10.jar:2.8.10]
	at io.druid.jackson.DefaultObjectMapper.<init>(DefaultObjectMapper.java:47) ~[druid-processing-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.jackson.DefaultObjectMapper.<init>(DefaultObjectMapper.java:35) ~[druid-processing-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.jackson.JacksonModule.jsonMapper(JacksonModule.java:46) ~[druid-processing-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.jackson.JacksonModule$$FastClassByGuice$$fa1bdb70.invoke(<generated>) ~[druid-processing-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.doProvision(ProviderMethod.java:264) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderMethod$Factory.provision(ProviderMethod.java:401) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderMethod$Factory.get(ProviderMethod.java:376) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1092) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:194) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:56) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingleParameterInjector.inject(SingleParameterInjector.java:38) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingleParameterInjector.getAll(SingleParameterInjector.java:62) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:110) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:90) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:268) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1092) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:194) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingleParameterInjector.inject(SingleParameterInjector.java:38) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingleParameterInjector.getAll(SingleParameterInjector.java:62) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.SingleMethodInjector.inject(SingleMethodInjector.java:82) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:132) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:93) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:80) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1085) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.MembersInjectorImpl.injectAndNotify(MembersInjectorImpl.java:80) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.Initializer$InjectableReference.get(Initializer.java:223) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.Initializer.injectAll(Initializer.java:132) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:174) ~[guice-4.1.0.jar:?]
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:110) ~[guice-4.1.0.jar:?]
	at com.google.inject.Guice.createInjector(Guice.java:99) ~[guice-4.1.0.jar:?]
	at com.google.inject.Guice.createInjector(Guice.java:73) ~[guice-4.1.0.jar:?]
	at io.druid.guice.GuiceInjectors.makeStartupInjector(GuiceInjectors.java:60) ~[druid-processing-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexer.HadoopDruidIndexerConfig.<clinit>(HadoopDruidIndexerConfig.java:106) ~[druid-indexing-hadoop-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at io.druid.indexing.common.task.HadoopIndexTask$HadoopDetermineConfigInnerProcessing.runTask(HadoopIndexTask.java:311) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at io.druid.indexing.common.task.HadoopTask.invokeForeignLoader(HadoopTask.java:228) ~[druid-indexing-service-0.13.0-SNAPSHOT.jar:0.13.0-SNAPSHOT]
	... 7 more
```

I guess the last error is because of the mismatched Jackson version when the peon tries to initialize some variables using Guice injection which means a newer version of Jackson is required. I'm not sure how the second and the third errors can be fixed (maybe by shading?).

I understand what @gianm and @b-slim are concerned with. I'm doing further testing to figure out what's the best way. I'll post the result here once my testing is done.

drcrallen commented:
@jihoonson does this deprecate https://github.com/druid-io/druid/blob/druid-0.12.0-rc1/pom.xml#L1170-L1176 the `spark2` profile?

drcrallen commented:
Related: https://github.com/druid-io/druid/pull/5288

drcrallen commented on extensions-core/s3-extensions/pom.xml:
```diff
@@ -46,16 +46,16 @@
-            <groupId>net.java.dev.jets3t</groupId>
-            <artifactId>jets3t</artifactId>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-s3</artifactId>
```
is this still needed since aws-common pulls in the bundle?

jihoonson replied:
Yes, maybe. I didn't test yet.

drcrallen commented:
Please add the removal of `AWSSessionCredentialsAdapter` to the master PR comment with justification (i.e. does the aws sdk do this already?)

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentMover.java:
```diff
@@ -155,40 +158,39 @@ private void selfCheckingMove(
+      final S3ObjectSummary objectSummary = listResult.getObjectSummaries().get(0);
+      if (objectSummary.getStorageClass() != null &&
+          StorageClass.fromValue(StringUtils.toUpperCase(objectSummary.getStorageClass())).equals(StorageClass.Glacier)) {
+        throw new ISE(
```
This is to keep it from being caught in a retry loop? can a comment be added to the code here as such?

jihoonson replied:
Looks like my fault. I don't want to change any logic in this PR. Changed to AmazonServiceException.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPuller.java:
```diff
@@ -89,18 +89,70 @@ public String getName()
+            @Override
+            public void close() throws IOException
+            {
+              delegate.close();
```
`s3Object` should still be closed even if delegate close has an error

jihoonson replied:
Thanks, fixed.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPuller.java:
```diff
@@ -89,18 +89,70 @@ public String getName()
-            return stream;
           }
+
+          return new InputStream()
```
would `FilterInputStream` with a `close` override work here?

jihoonson replied:
Done.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPusher.java:
```diff
@@ -163,21 +155,25 @@ public DataSegment push(final File indexFilesDir, final DataSegment inSegment, f
+      log.info("Skipping push because key [%s] exists && replaceExisting == false", key);
     } else {
-      s3Client.putObject(bucketName, object);
+      log.info("Pushing [%s] to bucket[%s] and key[%s].", file, bucket, key);
```
(nit-pick) would this be better if the `s3client` had a logger?

We're not really adding any information here that couldn't be gathered from reasonable logging in the `s3client` I don't think?

jihoonson replied:
I'm not sure. I just moved this log which was outside of the if clause to its inside.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPusher.java:
```diff
@@ -163,21 +155,25 @@ public DataSegment push(final File indexFilesDir, final DataSegment inSegment, f
+      s3Client.putObject(indexFilePutRequest);
     }
+
+    log.info("Deleting file [%s]", file.getAbsolutePath());
```
suggest adding in info as to why, like `Deleting temporary cached file [%s]`

jihoonson replied:
Done.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPusher.java:
```diff
@@ -163,21 +155,25 @@ public DataSegment push(final File indexFilesDir, final DataSegment inSegment, f
     }
+
+    log.info("Deleting file [%s]", file.getAbsolutePath());
+    file.delete();
```
Do we care if it was successful?

jihoonson replied:
Not sure. The original code didn't care about it.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3TimestampVersionedDataFinder.java:
```diff
@@ -65,23 +68,28 @@ public URI getLatestVersion(final URI uri, final @Nullable Pattern pattern)
+            while (objectSummaryIterator.hasNext()) {
+              final S3ObjectSummary objectSummary = objectSummaryIterator.next();
+              // TODO: what is going on here?
+              String keyString = objectSummary.getKey().substring(coords.path.length());
```
it is taking the path beyond the "search" prefix, and checking the regex to see if it matches the remainder (after the prefix)

jihoonson replied:
Thanks.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java:
```diff
@@ -88,121 +84,172 @@ public boolean apply(Throwable e)
   }
 
-  public static String constructSegmentPath(String baseKey, String storageDir)
+  static String constructSegmentPath(String baseKey, String storageDir)
```
we don't know if these were used in extensions, can the things that were public remain so?

jihoonson replied:
Reverted.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java:
```diff
@@ -88,121 +84,172 @@ public boolean apply(Throwable e)
+    if (key.endsWith("/") && objectMetadata.getContentLength() == 0) {
+      return true;
+    }
+    // Recognize s3sync.rb directory placeholders by MD5/ETag value.
```
is this string in source somewhere?

jihoonson replied:
Yes, it was in the original code.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java:
```diff
@@ -88,121 +84,172 @@ public boolean apply(Throwable e)
+    }
+    final S3ObjectSummary objectSummary = result.getObjectSummaries().get(0);
+    if (!objectSummary.getBucketName().equals(bucket) || !objectSummary.getKey().equals(key)) {
+      throw new ISE("Wrong object[%s] for bucket[%s] and key[%s]", objectSummary, bucket, key);
```
Does this mean that if there are multiple keys that match the prefix, then if the returned value isn't the first one in the result this will fail?  that seems kind of fragile.

jihoonson replied:
Yes, I added javadoc for this method.

jihoonson commented:
> @jihoonson does this deprecate https://github.com/druid-io/druid/blob/druid-0.12.0-rc1/pom.xml#L1170-L1176 the spark2 profile?

Yes, maybe. I didn't test yet.

> Please add the removal of AWSSessionCredentialsAdapter to the master PR comment with justification (i.e. does the aws sdk do this already?)

`AWSSessionCredentialsAdapter` was used by only `S3StorageDruidModule` to pass credentials from  aws-java-sdk's `AWSCredentialsProvider` to jet3t's `Rest3Client`. We don't need it anymore.

drcrallen commented on server/src/main/java/io/druid/server/initialization/jetty/JettyServerModule.java:
```diff
@@ -60,7 +60,7 @@
 import io.druid.server.metrics.MetricsModule;
 import io.druid.server.metrics.MonitorsConfig;
-import org.apache.http.HttpVersion;
+import io.netty.handler.codec.http.HttpVersion;
```
this is a strange change, is it intended?

jihoonson replied:
Yes, this module has no longer dependency on HttpClient, and the only usage of `HttpVersion` is to get the version string of HTTP 1.1 (`HttpVersion.HTTP_1_1.toString()`). To reduce confusion, I added a static string variable for HTTP 1.1 version string and removed this import.

jihoonson commented:
@drcrallen thanks! I'm doing further tests to support older versions of Hadoop (especially 2.7 line) as well as resolving my TODOs. I'll update this PR once it's ready.

gianm commented:
I am ok with this solution in principle. How would we do it though?

drcrallen commented:
What is the tolerance for suggesting hadoop 2.9 or higher?

gianm commented:
My sense is that Hadoop 2.9.0 was only released in Nov 2017 so it's too new to make the minimum version. People don't tend to update their Hadoop clusters very often.

jihoonson commented:
> I am ok with this solution in principle. How would we do it though?

We may fork the `hadoop-aws` module as a project under `druid-io`, and make some changes to support hadoop 2.7 line. And we may release the customized `hadoop-aws.jar` along with our next release (0.13). Does it make sense?

jihoonson commented:
@drcrallen would you please answer for my question [here](https://github.com/druid-io/druid/pull/2900/files#r155724414)?

drcrallen commented:
@jihoonson answered

gianm commented:
> @gianm We may fork the hadoop-aws module as a project under druid-io, and make some changes to support hadoop 2.7 line. And we may release the customized hadoop-aws.jar along with our next release (0.13). Does it make sense?

Will this work with a remote cluster? I thought - usually - the remote cluster prefers its own Hadoop jars rather than any Hadoop jars that user code (i.e. Druid) might ship. I guess it's fine since on the remote cluster, Druid probably won't be using aws-java-sdk directly, so there won't be an opportunity to have problems there?

b-slim commented:
@jihoonson thanks for doing this,  are you planning on shading/relocation as well? can you please elaborate more on what part will be forking if you have to?

jihoonson commented:
I've done some more tests and updated the PR description. One thing related to Hadoop support is that this PR works with Hadoop 2.7.5 if S3NFileSystem is set to use. The below is a quote from Hadoop Wiki.

> Unmainteained: S3N FileSystem (URI scheme: s3n://)
> S3N is the S3 Client for Hadoop 2.6 and earlier. From Hadoop 2.7+, switch to s3a
>
> A native filesystem for reading and writing regular files on S3.With this filesystem is that you can access files on S3 that were written with other tools. Conversely, other tools can access files written using Hadoop. The S3N code is stable and widely used, but is not adding any new features (which is why it remains stable).
>
> S3N requires a compatible version of the jets3t JAR on the classpath.
>
> Since Hadoop 2.6, all work on S3 integration has been with S3A. S3N is not maintained except for security risks —this helps guarantee security. Most bug reports against S3N will be closed as WONTFIX and the text "use S3A". Please switch to S3A if you can -and do try it before filing bug reports against S3N.

If only new features are not being added to S3N, I think it's fine. But, if there are some bugs, using S3N for Hadoop 2.7 won't work. @b-slim do you have any idea about this?

jihoonson commented:
I found that the current master branch also doesn't work with Hadoop 2.7.5 S3AFileSystem. I think this issue should be handled separately.

jihoonson commented:
S3A issue relates to #4841.

jihoonson commented:
Ready for review.

jihoonson commented on extensions-core/druid-kerberos/src/main/java/io/druid/security/kerberos/KerberosAuthenticator.java:
```diff
@@ -334,13 +330,14 @@ public Principal getUserPrincipal()
-                                 token.getExpires(),
-                                 !token.isExpired() && token.getExpires() > 0,
-                                 isHttps
+                createAuthCookie(
```
@b-slim the signature of `createAuthCookie()` was changed in 2.8. Would you check this change is valid?

b-slim replied:
Thought that copying the body of the function will avoid issues if we want to support multiple Hadoop versions. Is there any reason to not have the code living in Druid?

jihoonson replied:
Hmm, do you mean this break the compatibility with Hadoop 2.7? If so, would you elaborate on how it breaks? I didn't test, but looks like it would work because `hadoop-common.jar` is in `$DRUID_HOME/extensions/druid-kerberos/` and that `hadoop-common.jar` will be used once `druid-kerberos` extension is loaded.

b-slim replied:
having that function inline means we can compile druid with 2.7.X and 3.0.X.

jihoonson replied:
Hmm, do we need that? I think most users will use an already compiled druid rather than compiling by themselves. I think it would be fine if this works with 2.7. I'll test it.

b-slim replied:
@jihoonson at least am sure we do and probably other folks too...we compile against internal version to make sure that at runtime we have the same version everywhere

jihoonson replied:
Ah, all right. I'll revert this change.

b-slim commented on extensions-core/hdfs-storage/pom.xml:
```diff
@@ -152,10 +152,6 @@
             </exclusion>
           </exclusions>
         </dependency>
-        <dependency>
```
why is this removed, this will break the HDFS module.

b-slim replied:
@jihoonson FYI one of the use case of HDFS module is to load data from S3A using HDFS APIs thus the need for drivers.

jihoonson replied:
Why does this break the HDFS module? AFAIK, S3A is in `hadoop-aws.jar` which needs to be added to the extension classpath. Druid's hdfs-storage doesn't use any API of `aws-java-sdk`.

b-slim replied:
but i see this 
```
<exclusions>
            <exclusion>
              <groupId>com.amazonaws</groupId>
              <artifactId>aws-java-sdk</artifactId>
            </exclusion>
          </exclusions>
```
thus i think why it is needed...

b-slim replied:
i think what you are suggestion will lead to https://hadoop.apache.org/docs/stable/hadoop-aws/tools/hadoop-aws/index.html#ClassNotFoundException:_com.amazonaws.services.s3.AmazonS3Client

jihoonson replied:
I think it would be better to add a proper version of aws-java-sdk to the extension classpath and this is easy if `pull-deps` utility is used. It downloads all necessary dependencies at once. Forcing a fixed version of aws-java-sdk for various hadoop versions might cause many potential version conflicts.

b-slim replied:
@jihoonson pull-deps is not an option for security concerns, a lot of Druid cluster cannot access the internet for obvious security reasons. not sure what is the concern with having HDFS module packaged with the necessary dependencies.

jihoonson replied:
> not sure what is the concern with having HDFS module packaged with the necessary dependencies.

Having dependencies for hadoop-aws and aws-java-sdk here doesn't provide the necessary dependencies. It means that the hadoop-aws of `${hadoop.compile.version}` and aws-java-sdk of `${aws.sdk.bundle.version}` are included in the classpath. Obviously this can make potential conflicts with various hadoop versions. Even for secure hadoop clusters, they need to install hadoop-aws and aws-java-sdk in some ways even if that is not directly downloading from internet.

jihoonson replied:
Changed the scope of hadoop-aws to `provided` as well.

b-slim replied:
@jihoonson not sure what you mean by 
 >   Even for secure hadoop clusters, they need to install hadoop-aws and aws-java-sdk in some ways even if that is not directly downloading from internet.

Nothing needs to be installed if all the dependency is part of HDFS module directory. This is needed for historical as well. Same thing for Hadoop common it is part of the module.

jihoonson replied:
@b-slim probably my explanation was not detailed enough. Sorry for that.

As you know, `druid-hdfs-storage` module is an extension for using HDFS as deep storage for druid. To do so, druid calls some HDFS APIs which requires to include some Hadoop dependencies like `hadoop-common` and `hadoop-client` by default. 

However, `hadoop-aws` is somewhat different because it is used in HDFS side and is not needed in druid side. If users want to make their HDFS working with s3, they should have already installed `hadoop-aws` in somewhere when HDFS starts up. Then, they can add `hadoop-aws` to Druid's classpath easily (this is need to push/pull segments to/from HDFS). If they don't use HDFS but want to use s3, they're encouraged to use `druid-s3-extensions` rather than `druid-hdfs-storage`. 

`aws-java-sdk` is used in `hadoop-aws`, but not in `druid-hdfs-storage`. Since the version of `aws-java-sdk` is different according to which version of `hadoop-aws` is used, it should be provided along with `hadoop-aws`. If we fix the version of `aws-java-sdk` here, it might mismatch with the one used in `hadoop-aws` and cause operation failures when trying to access to s3.

What do you think?

b-slim replied:
@jihoonson i don't agree, i don't see why this should happen, i think druid-hdfs-storge should be able to package all its dependency and same for druid-s3-extensions, in the end only one extension is loaded thus i don't see how the conflict will happen unless you are telling me that Druid core class loader contains aws-java-sdk jars even if the user is not using druid-s3-extensions which is wrong IMO since Druid core does not need this to work. Does that make sense?

nishantmonu51 replied:
" in the end only one extension is loaded" - is not entirely correct, the historicals may want to load both the extensions and independent segments can be loaded from different deep storage. e.g. A user migrated the deep storage from hdfs to s3 but choose to keep older segments in hdfs only. 

However,  if both extensions load their dependencies in their own class loaders, there should not be any conflict.

b-slim replied:
@nishantmonu51 I don't think Druid allow you to load both at the same time since the is one property string that does the binding. Today if you load both you have a dependency injection issue. Unless I am missing something?

jihoonson replied:
Ok. First of all, I'm telling you that conflicts doesn't exist in druid, but exist when druid interacts with hadoop. In druid, there are two modules, `indexing-hadoop` and `hdfs-storage`. In both modules, druid calls some hdfs apis. Another thing which should be noted is that druid and hadoop have many dependencies on the same library but different versions. So, when druid calls some hdfs apis, the version of those libraries should be matched or at least compatible. If not, it causes conflicts.

> i think druid-hdfs-storge should be able to package all its dependency and same for druid-s3-extensions

@b-slim I agree on that we will be happy if we can do this, but I think, in this case, druid-hdfs-storage can't package all its dependencies because the version of those dependencies varies with different hadoop versions. If we can't achieve this, I think it would be better for users to put proper libraries because at least users can match the library versions by themselves. If the library version is fixed, there is no way to avoid conflicts.

> Druid core class loader contains aws-java-sdk jars even if the user is not using druid-s3-extensions

`aws-module` is one of Druid's core module and it has a dependency on `aws-java-sdk`. 

> FYI one of the use case of HDFS module is to load data from S3A using HDFS APIs thus the need for drivers.

Finally, I've tested the current master with hadoop 2.7.5 by using s3 as a deep storage via hdfs, and it doesn't work due to a removed API in aws-java-sdk. I've test both S3A and S3N. @b-slim is this use case one of your customers? Are they using open source druid and hadoop or some version of hortonworks? If it's the latter case, I think this patch should be fine.

b-slim replied:
@jihoonson I have a question why Druid core need to load AWS stuff if it does not run or interact with AWS this seems to be a poor choice?

jihoonson replied:
That's a good question. I'm also not sure why it is, but it was there since 2015. Maybe we can also change it to be provided when loading an extension. I think it should be done in a follow-up pr if needed.

drcrallen commented on aws-common/src/main/java/io/druid/common/aws/AWSEndpointConfig.java:
```diff
@@ -0,0 +1,52 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class AWSEndpointConfig
```
Why do we need to have this class instead of relying on https://github.com/aws/aws-sdk-java/blob/1.11.291/aws-java-sdk-core/src/main/java/com/amazonaws/SDKGlobalConfiguration.java ?

jihoonson replied:
Sorry, I'm not sure how to use that. Does `SDKGlobalConfiguration` provide an endpoint configuration?

drcrallen replied:
They are system properties the sdk picks up as per: https://github.com/aws/aws-sdk-java/blob/4734de6fb0f80fe5768a6587aad3b9d0eaec388f/aws-java-sdk-core/src/main/java/com/amazonaws/auth/SystemPropertiesCredentialsProvider.java#L35

drcrallen replied:
Basically if there's functionality in this config that is not available in the SDK supported system properties and env vars, then that is fine. But if all the config stuff can be done in SDK idiomatic ways, then I suggest we simply document where to find info on the SDK.

jihoonson replied:
@drcrallen the linked doc is for credentials and we're already using it for credentials.

AFAIK, aws-java-sdk doesn't provide such a configuration chain for the endpoint configuration. Instead, it's recommended to use `AwsClientBuilder.EndpointConfiguration` (https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/client/builder/AwsClientBuilder.EndpointConfiguration.html) or `AmazonS3Client.setEndpoint()` (https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/AmazonWebServiceClient.html#setEndpoint-java.lang.String-java.lang.String-java.lang.String-). I used the later one even though it's deprecated because using `AwsClientBuilder` causes some injection problem. I didn't dig that problem and not sure how to fix it.

drcrallen replied:
As another example: https://github.com/aws/aws-sdk-java/blob/1.11.293/aws-java-sdk-core/src/main/java/com/amazonaws/regions/LegacyRegionXmlMetadataBuilder.java allows specifying region endpoints. But those might be such an esoteric way of specifying things then it could be better to just handle it on our own until more obvious ways arise.

drcrallen commented on aws-common/src/main/java/io/druid/common/aws/AWSProxyConfig.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @JsonProperty
+  public String getPassword()
```
can this be a `PasswordProvider`?

jihoonson replied:
What do you mean? If you mean `AWSCredentialsProvider`, it can be used here because this is for proxy configuration. If you mean we need a sort of credentialsProvider for proxy, I agree but want to do in a follow-up pr.

drcrallen replied:
I mean a http://druid.io/docs/latest/operations/password-provider.html

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3StorageDruidModule.java:
```diff
@@ -102,15 +109,44 @@ public void configure(Binder binder)
+    if (StringUtils.isNotEmpty(proxyConfig.getHost())) {
+      conf.setProxyHost(proxyConfig.getHost());
+    }
+    if (proxyConfig.getPort() != -1) {
```
`-1` means "use default" in java URL world if I recall. does this need to be an invalid proxy number like `-2`?

jihoonson replied:
AWS's default proxy port is `-1`, so this means we will use the default configuration of AWS. I added a comment about this.

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3StorageDruidModule.java:
```diff
@@ -102,15 +109,44 @@ public void configure(Binder binder)
+    if (StringUtils.isNotEmpty(proxyConfig.getUsername())) {
+      conf.setProxyUsername(proxyConfig.getUsername());
+    }
+    if (StringUtils.isNotEmpty(proxyConfig.getPassword())) {
```
move to password provider please

drcrallen commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java:
```diff
@@ -88,91 +83,81 @@ public boolean apply(Throwable e)
-      }
-      if ("AccessDenied".equals(e.getErrorCode())) {
+    catch (AmazonS3Exception e) {
+      if (e.getStatusCode() == 403) {
```
is 404 not valid anymore?

jihoonson replied:
Good catch! Fixed it.

drcrallen commented on indexing-hadoop/src/main/java/io/druid/indexer/DetermineHashedPartitionsJob.java:
```diff
@@ -396,7 +396,8 @@ public void run(Context context)
 
-      if (config.get("mapred.job.tracker").equals("local") || determineIntervals) {
+      final String jobTrackerString = config.get("mapred.job.tracker");
+      if ((jobTrackerString != null && jobTrackerString.equals("local")) || determineIntervals) {
```
`"local".equals(config.get("mapred.job.tracker"))`

jihoonson replied:
Done.

drcrallen commented on processing/src/test/java/io/druid/query/topn/AlphaNumericTopNMetricSpecTest.java:
```diff
@@ -104,8 +104,8 @@ public void testSerdeAlphaNumericTopNMetricSpec() throws IOException
     ObjectMapper jsonMapper = TestHelper.makeJsonMapper();
-    TopNMetricSpec actualMetricSpec = jsonMapper.readValue(jsonMapper.writeValueAsString(jsonMapper.readValue(jsonSpec, TopNMetricSpec.class)), AlphaNumericTopNMetricSpec.class);
-    TopNMetricSpec actualMetricSpec1 = jsonMapper.readValue(jsonMapper.writeValueAsString(jsonMapper.readValue(jsonSpec1, TopNMetricSpec.class)), AlphaNumericTopNMetricSpec.class);
+    TopNMetricSpec actualMetricSpec = jsonMapper.readValue(jsonMapper.writeValueAsString(jsonMapper.readValue(jsonSpec, TopNMetricSpec.class)), TopNMetricSpec.class);
```
are all these test changes meant to be in this PR?

jihoonson replied:
Yes, it causes a syntax error due to the changed Jackson version.

drcrallen commented on server/src/main/java/io/druid/server/AsyncManagementForwardingServlet.java:
```diff
@@ -19,6 +19,7 @@
 
 package io.druid.server;
 
+import com.amazonaws.thirdparty.apache.http.client.utils.URIBuilder;
```
Should this really be using the shaded one?

jihoonson replied:
Nice catch! Fixed it.

drcrallen commented on server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java:
```diff
@@ -19,21 +19,22 @@
 
 package io.druid.server;
 
+import com.amazonaws.thirdparty.apache.http.client.utils.URIBuilder;
```
same question

jihoonson replied:
Fixed.

drcrallen commented on server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java:
```diff
@@ -444,13 +445,11 @@ public void onComplete(Result result)
                         "success",
                         success
-                        && result.getResponse().getStatus() == javax.ws.rs.core.Response.Status.OK.getStatusCode()
+                        && result.getResponse().getStatus() == Status.OK.getStatusCode()
```
are these changes part of this PR?

jihoonson replied:
Well, this is a simple change to make code look better.

drcrallen commented on server/src/main/java/io/druid/server/initialization/jetty/JettyServerModule.java:
```diff
@@ -269,7 +269,7 @@ static Server makeAndInitializeServer(
       final ServerConnector connector = new ServerConnector(
           server,
-          new SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.toString()),
+          new SslConnectionFactory(sslContextFactory, HTTP_1_1_STRING),
```
part of PR?

jihoonson replied:
https://github.com/druid-io/druid/pull/5382#discussion_r168253264

drcrallen commented on server/src/test/java/io/druid/query/dimension/LookupDimensionSpecTest.java:
```diff
@@ -70,12 +71,13 @@
     );
     String serLookup = mapper.writeValueAsString(lookupDimSpec);
-    Assert.assertEquals(lookupDimSpec, mapper.reader(LookupDimensionSpec.class).with(injectableValues).readValue(serLookup));
+    Assert.assertEquals(lookupDimSpec, mapper.reader(DimensionSpec.class).with(injectableValues).readValue(serLookup));
```
part of PR?

jihoonson replied:
Needed due to the changed Jackson version.

jihoonson commented:
@drcrallen any other comments?

drcrallen commented:
@jihoonson just https://github.com/druid-io/druid/pull/5382/files#r175774341 similar to https://github.com/druid-io/druid/blob/89215382510a6eceac898d14022b76a065fd7cdf/aws-common/src/main/java/io/druid/common/aws/AWSCredentialsConfig.java#L34

jihoonson commented:
@drcrallen ok, I agree it's needed. But, I would like to do in a follow-up PR because this PR easily conflicts with others. If you think its' fine, then I'll raise an issue for it. If not, I'll separate the proxy configuration from this PR and make another one. What do you think?

jihoonson commented:
@drcrallen thanks. Raised https://github.com/druid-io/druid/issues/5507.

jon-wei commented on extensions-core/s3-extensions/pom.xml:
```diff
@@ -45,17 +45,12 @@
-            <artifactId>jets3t</artifactId>
-            <scope>provided</scope>
-        </dependency>
+      <dependency>
```
indentation looks off here

jihoonson replied:
Fixed.

jon-wei commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPusher.java:
```diff
@@ -163,21 +155,29 @@ public DataSegment push(final File indexFilesDir, final DataSegment inSegment, f
     }
+
+    log.info("Deleting temporary cached file [%s]", file.getAbsolutePath());
+    file.delete();
```
Would it be better to clean this file up in a finally block?

jihoonson replied:
Good point. Fixed.

jon-wei commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java:
```diff
@@ -184,25 +169,93 @@ public static String constructSegmentPath(String baseKey, String storageDir)
+    if (result.getKeyCount() == 0) {
+      throw new ISE("Cannot find object for bucket[%s] and key[%s]", bucket, key);
+    }
+    final S3ObjectSummary objectSummary = result.getObjectSummaries().get(0);
```
Is it possible for result.getKeyCount() to be > 0 but result.getObjectSummaries() to have size 0?

jihoonson replied:
From the javadoc of [ListObjectsV2Result](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/ListObjectsV2Result.html), `getKeyCount()` returns number of keys returned with this response, and `getObjectSummaries()` returns a list of the object summaries describing the objects stored in the S3 bucket. Since each object associates with a key in S3, it shouldn't happen.

jon-wei commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java:
```diff
@@ -184,25 +169,93 @@ public static String constructSegmentPath(String baseKey, String storageDir)
     return filename;
   }
+
+  static AccessControlList grantFullControlToBucketOwver(AmazonS3 s3Client, String bucket)
```
Owver -> Owner

jihoonson replied:
Fixed.

jon-wei commented on extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3Utils.java:
```diff
@@ -184,25 +169,93 @@ public static String constructSegmentPath(String baseKey, String storageDir)
+
+  /**
+   * Gets a single {@link S3ObjectSummary} from s3. Since this method might throw an exception if there are multiple
+   * objets that match the given key, this method should be used only when it's guaranteed that the given key is unique
```
objets -> objects

jihoonson replied:
Fixed.

jihoonson commented:
@jon-wei thanks for the review. Addressed your comments.

jihoonson commented:
I'm going to merge this PR shortly. @b-slim @drcrallen @jon-wei thanks for the review!

b-slim commented:
@jihoonson i see this PR is merged so what is the story now? are we dropping support for Hadoop < 2.9 if, that is the case did we have a round of vote at mailing list? is this suppose to be 0.13.0 ?

jihoonson commented:
@b-slim we still support Hadoop 2.7 line. I tested Hadoop 2.7.3 and 2.7.5. When Hadoop 2.7 is used with s3 deep storage together, S3N still works but S3A doesn't work which is same for the master before this PR. To use S3A, they need a custom `hadoop-aws` which shades all conflicted libraries like aws, guava, and jackson.

jihoonson commented:
@b-slim 

> -First as per this thread we are suppose to have support from 2.6.0 and higher, https://groups.google.com/forum/#!topic/druid-development/vFjR0pdq83U. So am not sure if you have done testing with that as well or we are dropping support for any version less than 2.7.3?

I tested Apache Hadoop 2.6.0, 2.7.1, 2.7.3, 2.7.5, 2.8.3, 2.9.0, 3.0.0, Cloudera 5.7.0, and HDP 5.7. But, it would be great if you can test more Apache Hadoop and HDP versions as well.

> -Second, am wondering, have you tested that with HDFS as deep storage or only S3?

I've tested four cases when deep storage is HDFS, S3A via HDFS, S3N via HDFS, and direct S3.

> -Third if it works with 2.7.X why changing the version Hadoop to 2.8.3?

Well, at first, I wanted to use Hadoop 2.9 to use more recent libraries especially Jackson. But, it broke the support for Hadoop 2.7, so I needed to change it again. IMO, it doesn't make sense to use Hadoop 2.7.3 as our default Hadoop version because we don't support the whole features for that version like S3A.
